### ISSUE TITLE
RFE: Dockerfile to build from working directory

### ISSUE THREAD

**Issue by pierreprinetti** [ether/etherpad-lite#3657]
Currently, the Dockerfile builds the Docker image after pulling the Etherpad codebase from Github:

https://github.com/ether/etherpad-lite/blob/801d519165ef64bf5424ac395abfc838fa25ba9d/docker/Dockerfile#L28-L42

---

**I would find more convenient (and less surprising) if the Dockerfile built the Docker image from the code checked out in the local filesystem.**

In practice, I suggest replacing the `curl` call with (I simplify):

```Dockerfile
COPY ./ ./
```

Accidentally, this change would enable testing the Docker image as part of the CI jobs.

---

What do you think?

cc @muxator

**Comment by muxator** [comment]
Ack, let's do it.

I would like this:
1. by default, building without any option would build the current working copy, **including any eventual uncommitted diff** as you proposed;
2. keep the possibility of using `--build-arg ETHERPAD_VERSION=someversion` to build a specific version, as checked out from the repo (that is, with a _clean_ working copy, and without hitting the network)

Before [multi stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) this would have meant including git and its dependencies in the image, but now we can think of using a first stage for implementing 1 and 2, and perform a final minimal build afterwards.

```Dockerfile
# Build
FROM node:10-buster-slim AS build

# this will only be present in the build container
RUN apt update && apt install -y git

# this would probably fail because ".." is out of scope of the build context. Work on it.
# https://docs.docker.com/engine/reference/commandline/build/#extended-description
COPY ../.git /opt/etherpad-lite

# Checkout required version whether is done in idiomatic git.
# I live in mercurial and my life is simpler :)
#
# Respecting 1 and 2 would probably require more logic (i.e. a checkout vs a simple COPY)
RUN git checkout NODE_VERSION

RUN rm -rf /opt/etherpad_lite/.git

# Package
FROM node:10-buster-slim
COPY --from=build /opt/etherpad-lite /opt/etherpad-lite

# continue normally as in the current Dockerfile...
```

Any comments? Would you be ok working on this?

**Comment by pierreprinetti** [comment]
> ".." is out of scope of the build context
I’d move the Dockerfile to the root level of the repo, if it’s OK for you. I don’t really have a great alternative in mind.

What is the need behind ETHERPAD_VERSION?

I find it simpler to keep things separate, i.e.

git checkout $revision && docker build .

... but I might be missing the point?

**Comment by pierreprinetti** [comment]
I have created a PR as a proof-of-concept: #3662 

With this simpler setting as a base, it would be even easier to set up an automated Docker tags for tagged versions.

**Comment by pierreprinetti** [comment]
If building Docker images out of old versions of Etherpad is still a requirement, we could keep a `legacy.Dockerfile` and document how to call it:
```
docker build -f legacy.Dockerfile -e ETHERPAD_VERSION=xxxyyyzzz -t etherpad .
```

**Comment by muxator** [comment]
Seeing the Dockerfile downloading a .tar.gz from Github every time a container is built is too ugly to keep it alive.
Let's go on with #3662 (see my comments there).

Later on, I would probably write an `alternative.Dockerfile`, because I still want a way to:
1. being sure that the code does not have any modifications wrt the committed version
2. use the _current_ version of the automation for building an _arbitrary_ version of the application, while guaranteeing 1)

While point 2) is a nitpick, point 1) seems important to me, and the version we are going to implement could only ensure this condition if the user cooperates working on a fresh clone. This is too an easy win: no invariant guarantees from the code :)

**Comment by pierreprinetti** [comment]
> #3662 (see my comments there).

I can't see them! :sob:

**Comment by muxator** [comment]
I think I forgot to click a green button :)
See if you see them now!

**Comment by muxator** [comment]
I had a hard time force pushing on https://github.com/pierreprinetti/etherpad/tree/local_dockerfile in order to bring #3662 in (my bad).

Directly implemented the rewritten version in dc15f4a43c0e.
Closing this.
